### PR TITLE
Decoder-Only Backbone

### PR DESCRIPTION
**Pull request by frinkleko**
**Todo:**
+ [x] Rebase from the code clean branch and add the decoder only backbone support.
+ [ ] Generate new IDF values for decoder only model.

### CODE REVIEW

**Comment by frinkleko** [comment]
Currently, we have implemented the data parallel, however, bug (model forward return nan) exists when we do distributed data parallel.

**Comment by frinkleko** [comment]
Rebase from the code clean branch and add the decoder only backbone support only.

File: scripts/model/sparse_encoders.py [hunk 1]
```diff
@@ -17,47 +30,77 @@ def __init__(
...
-        self.tokenizer = transformers.AutoTokenizer.from_pretrained(tokenizer_id)
-        self.special_token_ids = [
-            self.tokenizer.vocab[token]
-            for token in self.tokenizer.special_tokens_map.values()
-        ]
-
-        idf_vector = [1.0] * self.tokenizer.vocab_size
+
+        if model_type == "flan-t5":
+            self.backbone = T5EncoderModel.from_pretrained(model_id)
```
Comment (zhichao-aws) [inline scripts/model/sparse_encoders.py]: We're only using the encoder currently, but we should leverage its decoder to generate the logits on tokens

Comment (zhichao-aws) [reply scripts/model/sparse_encoders.py]: Here is an example to use decoder: https://github.com/huggingface/transformers/blob/2ab7bdc40333b230b642f09e8334fb8e1a92d2a4/src/transformers/models/t5/modeling_t5.py#L1698

For our use case there can be difference, we don't need to do greedy decode

Comment (frinkleko) [reply scripts/model/sparse_encoders.py]: I believe here is another [good reference](https://github.com/huggingface/transformers/issues/22241), while I encounter some `vocab_size` mismatch issue ([similar issues](https://github.com/huggingface/transformers/issues/4875)), let me work on it.
